### PR TITLE
Favorite backend

### PR DESCRIPTION
**Pull request by usmanmuhd**
##### Description

Adds backend for the favorites API

##### Refers/Fixes

Fixes: #4706 

##### Testing

No tests present currently.

CC: @octavioamu @danlipert

### CODE REVIEW

**Comment by danlipert** [comment]
@usmanmuhd Is this ready for review? If not, please use the "draft pull request" feature here on Github or just add [WIP] to the title of the PR 👍

**Comment by octavioamu** [comment]
@usmanmuhd how is going with this one? we need this ASAP

**Comment by kuhnchris** [comment]
@usmanmuhd asked me to take over this, so I'm on it today

**Comment by kuhnchris** [comment]
This now has a subdependency on #4749 - price/price_finney causes problems with the serializer

**Comment by kuhnchris** [comment]
@usmanmuhd the PR is ready as far as I can tell - GET, POST and DELETE works for /favorites/bounty, kudos, grants and user.
@octavioamu as discussed: please review. :-)

**Comment by thelostone-mc** [comment]
Could we get the travis to pass ?

**Comment by kuhnchris** [comment]
@thelostone-mc I asked @octavioamu he said he didn't want them squished. I can squish them no problemo.

**Comment by kuhnchris** [comment]
fyi the build failed thanks to slack... we may want to remove that check...

**Comment by octavioamu** [comment]
@kuhnchris I don't think the problem is with slack https://travis-ci.org/gitcoinco/web/builds/568469967#L2754
could be related to another pr merged

File: app/favorites/views.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+from rest_framework.decorators import api_view, authentication_classes, permission_classes
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from .models import Favorite
+
+
+@api_view(['POST', 'DELETE'])
+@authentication_classes((SessionAuthentication, BasicAuthentication))
```
Comment (octavioamu) [inline app/favorites/views.py]: @kuhnchris I don't think we are sharing basicauth and sessinauth with the front end just a http cookies, that means there is not possible to get http cookies from javascript. 
That is the way if you do a jwt auth or something like that and store the tokens on login to use later in any auth request.

Comment (octavioamu) [reply app/favorites/views.py]: unless is taking from the request tokens, I think that is the case, anyway get works perfectly but post give me 403

Comment (octavioamu) [reply app/favorites/views.py]: image

File: app/favorites/views.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        elif reqTypes[favType] == "GRAN":
+            for obj in Grant.objects.filter(id=fav.obj_id):
+                ser = GrantSerializer(obj)
+                jsonRetData.append(ser.data)
+        elif reqTypes[favType] == "USER":
+            for obj in Profile.objects.filter(user=fav.obj_id):
+                ser = ProfileSerializer(obj)
+                jsonRetData.append(ser.data)
+
+    return Response(json.dumps(jsonRetData, indent=4, sort_keys=True, default=str))
```
Comment (octavioamu) [inline app/favorites/views.py]: The responses are not json formatted 
image

Comment (octavioamu) [reply app/favorites/views.py]: return Response(**json.loads**(json.dumps(jsonRetData, indent=4, sort_keys=True, default=str)))
can solve it

File: app/favorites/views.py (hunk 1 above)
Comment (octavioamu) [reply app/favorites/views.py]: you can test it on the console with something like:
```
var postFav = fetchData('api/v0.1/favorites/user/2', 'POST')
$.when(postFav).then((response, status, statusCode) => {
  console.log(response, status, statusCode)
})
```

Comment (kuhnchris) [reply app/favorites/views.py]: Well if you are authorized (e.a. you have access to admin) then it does indeed work. (as i tried it myself) - I'm not 100% sure how we handle the github auth - do we have any other paths where we require authentification where I could steal the code from? Else we may need to check how to protect the data. (I tested it via Boomerang REST Extension for Chrome)

File: app/favorites/views.py (hunk 2 above)
Comment (kuhnchris) [reply app/favorites/views.py]: My bad. I had to play around with recursion and avoiding errors that I probabbly forgot about that. I'll add your fix in

File: app/favorites/views.py (hunk 1 above)
Comment (kuhnchris) [reply app/favorites/views.py]: Also, why would I use Javascript for auth? that's a pyhton file you are looking at - it checks on the backend (django) if you are auth - and if you are authed on the entire app it checks the User Object's permission and executes the function if the permission is sufficient. I do not think there are any additional API requirements, but we'd need to check RESTful API's documentation (django-rest-framework) if they may impose a special permissions...

Comment (kuhnchris) [reply app/favorites/views.py]: After a quick check - not sure but maybe I can use the django.contrib.auth.decorators.login_required.

It SHOULD provide a request.user if it does, and a request.user.profile (a gitcoin Profile object) if it is authenticated.

I can try that out if it's fine with you.

See dashboard/views.py and dashboard/utils.py - helper_handle_access_token and new_bounty method, those are taking those assumtions.

```
from django.contrib.auth.decorators import login_required
...
@login_required
def new_bounty(request):
...
        user=request.user if request.user.is_authenticated else None,
...
```

Comment (kuhnchris) [reply app/favorites/views.py]: OK I got what you mean now - i do get a 403, but the 403 tells me I need the CSRF token set. We do not have access to the CSRF Token tho. This only happens if we use SessionAuth. Which we do. So we either need to CSRFExempt the API calls (apparently not really viable, or easy doable), or use a different auth scheme like tokens, as you mentioned before.

Comment (octavioamu) [reply app/favorites/views.py]: yes but anyway try to comment the auth part and do a post with the command I sent to you.  Get are working with the same auth validator but this endpoint with Post doesn't work also without that

Comment (octavioamu) [reply app/favorites/views.py]: @kuhnchris how is going? any problem? Ping me if you have any question about my comment.

Comment (kuhnchris) [reply app/favorites/views.py]: well, all my attempts to implement this right now failed, from trying to bypass csrf to try to use another scheme (which we cannot) to trying to bypass the csrf somehow again. up to now success until now. any ideas/pointers? The only "idea" i could get at this point is to redo everything in a token-auth based thing but that'd break explorer and co. :-(

**Comment by thelostone-mc** [comment]
Closing this out as this feature ended up not being on the things we'd wanna build out ! 
Thanks @usmanmuhd for the PR  :D

**Comment by octavioamu** [comment]
Reopening seems we want to use this for townsquare favorite posts #6121

**Comment by octavioamu** [comment]
Reopening seems we want to use this for townsquare favorite posts #6121

**Pull request closed without merging**